File: apps/ios/modules/nearby/ios/ValenceNearbyModule.swift
import ExpoModulesCore
import Network

/// Browses the local network for Valence servers announcing themselves over
/// Bonjour, and says where each one answers.
///
/// An announcement names a service rather than an address, so each one is
/// connected to once, just long enough for the system to resolve it, and the
/// address it resolved to is what is reported. Only IPv4 is asked for, as the
/// desktop app only offers IPv4, since that is what somebody would type.
public class ValenceNearbyModule: Module {
  private var browser: NWBrowser?
  private var resolving: [String: NWConnection] = [:]
  private var found: [String: [String: Any]] = [:]

  public func definition() -> ModuleDefinition {
    Name("ValenceNearby")

    Events("onChange")

    Function("start") { (type: String) in
      self.start(type)
    }

    Function("stop") {
      self.stop()
    }

    OnDestroy {
      self.stop()
    }
  }

  /// Starts browsing for one kind of service, forgetting anything found before.
  private func start(_ type: String) {
    stop()

    let browser = NWBrowser(for: .bonjour(type: "_\(type)._tcp", domain: nil), using: NWParameters())

    browser.browseResultsChangedHandler = { [weak self] results, _ in
      self?.heard(results)
    }
    browser.start(queue: .main)
    self.browser = browser
  }

  /// Stops browsing and resolving, and forgets what was found.
  private func stop() {
    browser?.cancel()
    browser = nil
    resolving.values.forEach { $0.cancel() }
    resolving.removeAll()
    found.removeAll()
  }

  /// Takes what the browser can currently see: forgets what has gone and
  /// resolves what is new.
  private func heard(_ results: Set<NWBrowser.Result>) {
    let named = Dictionary(
      results.compactMap { result in nameOf(result.endpoint).map { ($0, result.endpoint) } },
      uniquingKeysWith: { first, _ in first }
    )

    for gone in found.keys where named[gone] == nil {
      found.removeValue(forKey: gone)
    }

    for (name, endpoint) in named where found[name] == nil && resolving[name] == nil {
      resolve(name, endpoint)
    }

    tell()
  }

  /// Connects to one announced service just long enough to learn its address.
  private func resolve(_ name: String, _ endpoint: NWEndpoint) {
    let parameters = NWParameters.tcp

    if let ip = parameters.defaultProtocolStack.internetProtocol as? NWProtocolIP.Options {
      ip.version = .v4
    }

    let connection = NWConnection(to: endpoint, using: parameters)

    connection.stateUpdateHandler = { [weak self, weak connection] state in
      guard let self, let connection else {
        return
      }

      switch state {
      case .ready:
        if case let .hostPort(host, port)? = connection.currentPath?.remoteEndpoint,
           case let .ipv4(address) = host {
          self.found[name] = [
            "name": name,
            "host": "\(address)".components(separatedBy: "%")[0],
            "port": Int(port.rawValue),
          ]
        }

        connection.cancel()
        self.resolving.removeValue(forKey: name)
        self.tell()
      case .failed, .cancelled:
        self.resolving.removeValue(forKey: name)
      default:
        break
      }
    }

    resolving[name] = connection
    connection.start(queue: .main)
  }

  /// The name a service announced itself under.
  private func nameOf(_ endpoint: NWEndpoint) -> String? {
    if case let .service(name, _, _, _) = endpoint {
      return name
    }

    return nil
  }

  /// Tells the app what is on the network now.
  private func tell() {
    sendEvent("onChange", ["nearby": Array(found.values)])
  }
}

File: apps/mobile/modules/nearby/android/src/main/java/app/valence/modules/nearby/ValenceNearbyModule.kt
package app.valence.modules.nearby

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Build
import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.Inet4Address

/**
 * Browses the local network for Valence servers announcing themselves over DNS-SD, and says where
 * each one answers, as the Apple side does with Bonjour.
 *
 * An announcement names a service rather than an address, so each one is resolved, one at a time
 * since older Androids refuse a second resolve while one is running, and only its IPv4 address is
 * reported, as that is what somebody would type.
 */
class ValenceNearbyModule : Module() {
  private val main = Handler(Looper.getMainLooper())
  private var manager: NsdManager? = null
  private var listening: NsdManager.DiscoveryListener? = null
  private val found = mutableMapOf<String, Map<String, Any>>()
  private val waiting = ArrayDeque<NsdServiceInfo>()
  private var isResolving = false

  override fun definition() = ModuleDefinition {
    Name("ValenceNearby")

    Events("onChange")

    Function("start") { type: String ->
      main.post { start(type) }
    }

    Function("stop") {
      main.post { stop() }
    }

    OnDestroy {
      main.post { stop() }
    }
  }

  /** Starts browsing for one kind of service, forgetting anything found before. */
  private fun start(type: String) {
    stop()

    val context = appContext.reactContext ?: return
    val nsd = context.getSystemService(Context.NSD_SERVICE) as NsdManager
    val discovery = object : NsdManager.DiscoveryListener {
      override fun onDiscoveryStarted(serviceType: String) = Unit

      override fun onDiscoveryStopped(serviceType: String) = Unit

      override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) = Unit

      override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) = Unit

      override fun onServiceFound(service: NsdServiceInfo) {
        main.post { queue(service) }
      }

      override fun onServiceLost(service: NsdServiceInfo) {
        main.post {
          found.remove(service.serviceName)
          tell()
        }
      }
    }

    manager = nsd
    listening = discovery
    nsd.discoverServices("_$type._tcp", NsdManager.PROTOCOL_DNS_SD, discovery)
  }

  /** Stops browsing, and forgets what was found. */
  private fun stop() {
    listening?.let { discovery -> runCatching { manager?.stopServiceDiscovery(discovery) } }
    listening = null
    manager = null
    waiting.clear()
    isResolving = false
    found.clear()
  }

  /** Puts a newly heard service in line to be resolved, unless it already has been. */
  private fun queue(service: NsdServiceInfo) {
    if (found.containsKey(service.serviceName) || waiting.any { it.serviceName == service.serviceName }) {
      return
    }

    waiting.addLast(service)
    resolveNext()
  }

  /** Resolves the next service in line, where none is being resolved already. */
  @Suppress("DEPRECATION")
  private fun resolveNext() {
    val nsd = manager ?: return

    if (isResolving) {
      return
    }

    val service = waiting.removeFirstOrNull() ?: return

    isResolving = true
    nsd.resolveService(service, object : NsdManager.ResolveListener {
      override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
        main.post {
          isResolving = false
          resolveNext()
        }
      }

      override fun onServiceResolved(serviceInfo: NsdServiceInfo) {
        main.post {
          isResolving = false
          addressOf(serviceInfo)?.let { host ->
            found[serviceInfo.serviceName] =
              mapOf("name" to serviceInfo.serviceName, "host" to host, "port" to serviceInfo.port)
            tell()
          }
          resolveNext()
        }
      }
    })
  }

  /** The IPv4 address a resolved service answers on. */
  @Suppress("DEPRECATION")
  private fun addressOf(service: NsdServiceInfo): String? {
    val addresses =
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) service.hostAddresses
      else listOfNotNull(service.host)

    return addresses.firstOrNull { it is Inet4Address }?.hostAddress
  }

  /** Tells the app what is on the network now. */
  private fun tell() {
    sendEvent("onChange", mapOf("nearby" to found.values.toList()))
  }
}
